exporter/loadbalancingexporter/resolver_k8s.go

```diff
@@ -0,0 +1,301 @@
+
+var (
+	errNoSvc                        = errors.New("no service specified to resolve the backends")
+	errNoSvcPorts                   = errors.New("no ports specified to resolve the backends")
+	k8sResolverMutator              = tag.Upsert(tag.MustNewKey("resolver"), "k8s service")
+	k8sResolverSuccessTrueMutators  = []tag.Mutator{resolverMutator, successTrueMutator}
+	k8sResolverSuccessFalseMutators = []tag.Mutator{resolverMutator, successFalseMutator}
+)
+
+type k8sResolver struct {
+	logger    *zap.Logger
+	service   string
+	namespace string
+	port      []int32
+
+	handler        *handler
+	once           *sync.Once
+	epsListWatcher cache.ListerWatcher
+	endpointsStore *sync.Map
+
+	endpoints         []string
+	onChangeCallbacks []func([]string)
+
+	stopCh             chan (struct{})
+	updateLock         sync.Mutex
+	shutdownWg         sync.WaitGroup
+	changeCallbackLock sync.RWMutex
+}
+
+func newK8sResolver(clt kubernetes.Interface,
+	logger *zap.Logger,
+	service string,
+	ports []int32) (*k8sResolver, error) {
+
+	if len(service) == 0 {
+		return nil, errNoSvc
+	}
+	if len(ports) == 0 {
+		return nil, errNoSvcPorts
+	}
+
+	nAddr := strings.SplitN(service, ".", 2)
+	name, namespace := nAddr[0], "default"
+	if len(nAddr) > 1 {
+		namespace = nAddr[1]
+	}
+
+	epsSelector := fmt.Sprintf("metadata.name=%s", name)
+	epsListWatcher := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			options.FieldSelector = epsSelector
+			return clt.CoreV1().Endpoints(namespace).List(context.TODO(), options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			options.FieldSelector = epsSelector
+			return clt.CoreV1().Endpoints(namespace).Watch(context.TODO(), options)
+		},
+	}
+
+	epsStore := &sync.Map{}
+	h := &handler{endpoints: epsStore, logger: logger}
+	r := &k8sResolver{
+		logger:         logger,
+		service:        name,
+		namespace:      namespace,
+		port:           ports,
+		once:           &sync.Once{},
+		endpointsStore: epsStore,
+		epsListWatcher: epsListWatcher,
+		handler:        h,
+		stopCh:         make(chan struct{}),
+	}
+	h.callback = r.resolve
+
+	return r, nil
+}
+
+func (r *k8sResolver) start(ctx context.Context) error {
+	if r.once == nil {
+		return fmt.Errorf("cannot Start() partial k8sResolver (nil *sync.Once)")
+	}
+	if r.handler == nil {
+		return fmt.Errorf("cannot Start() partial k8sResolver (nil *handler)")
+	}
+	var initErr error
+	r.once.Do(func() {
+		if r.epsListWatcher != nil {
+			r.logger.Debug("creating and starting endpoints informer")
+			epsInformer := cache.NewSharedInformer(r.epsListWatcher, &corev1.Endpoints{}, 0)
+			if _, err := epsInformer.AddEventHandler(r.handler); err != nil {
+				r.logger.Error("error adding event handler to endpoints informer", zap.Error(err))
+			}
+			go epsInformer.Run(r.stopCh)
+			if !cache.WaitForCacheSync(r.stopCh, epsInformer.HasSynced) {
+				r.logger.Error("error endpoints informer not sync")
```

Comment:
Same

Reply:
I have no idea what this means and what's the implication. Does this mean that the resolver will not work? Or will it work, just not with the cache?

Reply:
If the cache is not synced, the ListWatch on the endpoint resources will not work properly. So, when this happens, the `start` method here returns an error

Reply:
ok, I will remove log line.